src/codegen/c.zig

```diff
@@ -6288,28 +6334,42 @@ fn airMemset(f: *Function, inst: Air.Inst.Index, safety: bool) !CValue {
         }
         try writer.writeAll("; ++");
         try f.writeCValue(writer, index, .Other);
-        try writer.writeAll(") ((");
-        try f.renderType(writer, elem_ptr_ty);
-        try writer.writeByte(')');
-        try writeSliceOrPtr(f, writer, dest_slice, dest_ty);
-        try writer.writeAll(")[");
-        try f.writeCValue(writer, index, .Other);
-        try writer.writeAll("] = ");
-        try f.writeCValue(writer, value, .FunctionArgument);
-        try writer.writeAll(";\n");
+        try writer.writeAll(") ");
+        if (lowersToArray(elem_ty, target)) {
+            // Arrays are not assignable, so we use memcpy here.
+            try writer.writeAll("memcpy(");
+            try writeSliceOrPtr(f, writer, dest_slice, dest_ty);
+            try writer.writeAll("[");
+            try f.writeCValue(writer, index, .Other);
+            try writer.writeAll("], ");
+            try f.writeCValue(writer, value, .FunctionArgument);
+            try writer.print(", {d});\n", .{elem_abi_size});
+        } else {
+            try writer.writeAll("((");
+            try f.renderType(writer, elem_ptr_ty);
+            try writer.writeByte(')');
+            try writeSliceOrPtr(f, writer, dest_slice, dest_ty);
+            try writer.writeAll(")[");
+            try f.writeCValue(writer, index, .Other);
+            try writer.writeAll("] = ");
+            try f.writeCValue(writer, value, .FunctionArgument);
+            try writer.writeAll(";\n");
+        }
```

Comment:
```suggestion
        const a = try Assignment.start(f, writer, elem_ty);
        try writer.writeAll("((");
        try f.renderType(writer, elem_ptr_ty);
        try writer.writeByte(')');
        try writeSliceOrPtr(f, writer, dest_slice, dest_ty);
        try writer.writeAll(")[");
        try f.writeCValue(writer, index, .Other);
        try writer.writeByte(']');
        try a.assign(f, writer);
        try f.writeCValue(writer, value, .Other);
        try a.end(f, writer);
```

Reply:
I recently factored out these maybe-`memcpy`s.

Reply:
Beautiful, thank you!